### ISSUE TITLE
LoRA real run: one full epoch, then merge

### ISSUE THREAD

**Issue by kreuzhofer** [kreuzhofer/nebius-slurm-qwen38-demo#11]
## Question

What does a full LoRA epoch on this cluster actually cost and produce?

The real LoRA run: defaults untouched — one epoch, **583 steps** (74,649 examples at effective batch 128, measured from the downloaded dataset in #2), `LEARNING_RATE=2e-4`, `LORA_R=32`, `MAX_SEQ_LEN=1024` — followed by `merge_lora.py`. Every code path it uses has already been proven at smoke scale by this point, so this ticket is about cost and outcome, not about whether it works.

Record:

- Total wall-clock and tokens/s per GPU, and how that compares to the smoke run's rate (packing is unavailable for Gated DeltaNet, so the figure is expected to be lower than a dense model of this size)
- Final and best training loss, and the eval loss at whatever step evaluation lands on
- Peak memory per GPU
- On-disk size of adapter and merged checkpoint, and `/mnt/data` free space afterwards — the full-parameter path still needs ~54 GB of headroom after this

Before starting: confirm free space is enough for the merged checkpoint, and that no other job is on the nodes.

The accuracy number for this artifact is not measured here — it belongs to the three-way published evaluation, which is still fog until this run exists.

**Comment by kreuzhofer** [comment]
## Resolution

**The real LoRA run is done and the pipeline's canonical artifact exists.** Job 835, one full epoch on the settled configuration.

```
Targets      : attn+MLP (GDN dropped)   -- the config #12 settled on
trainable    : 159,383,552 (0.5891%)
Steps        : 1 epoch, no cap -> 584 steps
Train/eval   : 74,648 / 3,929 (full eval split)
train_runtime: 907.5 s   (82.25 samples/s, 0.644 steps/s)
loss         : 0.2647 -> 0.01437     eval_loss: 0.01691
Peak GPU mem : 29.08 GiB allocated, 63.83 GiB reserved  (rank 0)
merge        : 4m41s -> output/qwen3.8-27b-sql (51 GB, 2 shards)
errors       : 0     recovery fallback: silent
```

Disk: **270 GB free** after the merge. Adapter 4.3 GB (includes one retained step checkpoint), merged checkpoint 51 GB.

**`output/qwen3.8-27b-sql` is restored** — the documented zero-argument default for `evaluate.sbatch`, `serve.sbatch` and `query.sh`, which the smoke-artifact cleanup had removed. It now points at a real one-epoch model rather than a 25-step smoke adapter, which is what it should always have been.

### The memory gap is closed, and the README was wrong

This was flagged as "not captured" since [#9](https://github.com/kreuzhofer/nebius-slurm-qwen38-lora-demo/issues/9) and never fixed. Both training scripts now print `torch.cuda.max_memory_allocated/reserved` at the end of `main()`.

Measured: **29.08 GiB allocated, 63.83 GiB reserved**. LoRA holds only ~3.2 GiB of sharded base weights plus a negligible adapter, so **~25.6 GiB of that is activations and workspace** at `PER_DEVICE_BATCH_SIZE=8` / `MAX_SEQ_LEN=1024` with activation checkpointing on.

**That makes the README's "~20 GiB/GPU" for full fine-tuning a floor, not a total.** It counted parameter, gradient and optimizer states only. Full FT pays the same ~25.6 GiB activation cost on top, so project **~46 GiB/GPU allocated — about 17% of a 268.6 GiB card**. Still ample headroom; more than double the states-only figure. [#19](https://github.com/kreuzhofer/nebius-slurm-qwen38-lora-demo/issues/19) now has a number to test against instead of an untested claim. Corrected in the README (`1b31179`).

Worth carrying forward: three different memory numbers describe this one run — **allocated 29.08 GiB, reserved 63.83 GiB, and `nvidia-smi` ~50 GiB average with 68 GiB peaks**. Anyone comparing a prediction against a measurement needs to say which they mean.

### Timing, against the smoke-scale projections

907.5 s for 584 steps = 1.55 s/step average, versus the 1.04 s/step arm B managed at 500 steps. The difference is this run's two evaluations over the **full 3,929-example** eval split (the A/B runs also used the default, but this one's evals fell at step 500 *and* the epoch end) plus the checkpoint at step 500 and the final save. Pure training remains ~1.0-1.2 s/step. The [#9](https://github.com/kreuzhofer/nebius-slurm-qwen38-lora-demo/issues/9) estimate of "15-20 minutes for a full epoch" was accurate.

### Not measured here, by design

Accuracy for this artifact. It belongs to the three-way publication alongside the full fine-tune, which is still fog. The 500-step arm scored 89% and this run trained 584 steps on the same configuration, so expect something close — but that is an expectation, not a measurement, and it is not going in the README until it is one.
